source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,1896 @@
+warrant a new minor version. Small bug fixes and internal changes (e.g. grammar)
+will warrant a new patch version.
+
+Each test file defines a `schemaVersion`_, which test runners will use to
+determine compatibility (i.e. whether and how the test file will be
+interpreted). Test runners MAY support multiple versions of the test format.
+Test runners MUST NOT process incompatible files but are otherwise free to
+determine how to handle such files (e.g. skip and log a notice, fail and raise
+an error).
+
+Each major version of this specification will have a corresponding JSON schema
+(e.g. `schema-1.json <schema-1.json>`__), which may be used to programmatically
+validate YAML and JSON files using a tool such as `Ajv <https://ajv.js.org/>`__.
+
+The latest JSON schema MUST remain consistent with the `Test Format`_ section.
+If and when a new major version is introduced, the `Breaking Changes`_ section
+must be updated and JSON schema(s) for any previous major version(s) MUST remain
+available so that older test files can still be validated. New tests files
+SHOULD always be written using the latest version of this specification.
+
+
+Entity Map
+----------
+
+The entity map indexes arbitrary objects and values by unique names, so that
+they can be referenced from test constructs (e.g.
+`operation.object <operation_object_>`_). To ensure each test is executed in
+isolation, test runners MUST NOT share entity maps between tests. Most entities
+will be driver objects created by the `createEntities`_ directive during test
+setup, but the entity map may also be modified during test execution via the
+`operation.saveResultAsEntity <operation_saveResultAsEntity_>`_ directive.
+
+Test runners MAY choose to implement the entity map in a fashion most suited to
+their language, but implementations MUST enforce both uniqueness of entity names
+and referential integrity when fetching an entity. Test runners MUST raise an
+error if an attempt is made to store an entity with a name that already exists
+in the map and MUST raise an error if an entity is not found for a name or is
+found but has an unexpected type.
+
+Consider the following examples::
+
+    # Error due to a duplicate name (client0 was already defined)
+    createEntities:
+      - client: { id: client0 }
+      - client: { id: client0 }
+
+    # Error due to a missing entity (client1 is undefined)
+    createEntities:
+      - client: { id: client0 }
+      - session: { id: session0, client: client1 }
+
+    # Error due to an unexpected entity type (session instead of client)
+    createEntities:
+      - client: { id: client0 }
+      - session: { id: session0, client: client0 }
+      - session: { id: session1, client: session0 }
+
+
+Test Format
+-----------
+
+Each specification test file can define one or more tests, which inherit some
+top-level configuration (e.g. namespace, initial data). YAML and JSON test files
```

Comment:
I find it annoying that PRs always contain YAML and JSON versions.  Perhaps we should only review and commit the YAML, and generate the JSON when needed.

Reply:
I refer to "YAML and JSON" here because some drivers (e.g. Ruby) use the YAML files directly and have no need of the JSON files. My understanding is that we produce JSON as a convenience for drivers, since languages are more likely to have a full-featured JSON parser than a YAML parser.

I often use the PR's file filter to ignore `*.json` files, which removes the noise when committing. I know folks have discussed ideas for deferring JSON generation until merge-time, but nothing has come of that. That said, that sounds like something to be sorted out at the team/repository level and not in this PR.

Reply:
OK, fine to defer.